sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala

```diff
@@ -276,6 +368,77 @@ class TransformWithStateSuite extends StateStoreMetricsTest
       )
     }
   }
+
+  test("transformWithState - streaming with rocksdb and event time timer " +
+    "should succeed") {
+    withSQLConf(SQLConf.STATE_STORE_PROVIDER_CLASS.key ->
+      classOf[RocksDBStateStoreProvider].getName) {
+
+      import java.sql.Timestamp
+      val inputData = MemoryStream[(String, Timestamp)]
+      val result = inputData.toDS()
+        .select($"_1".as("value"), $"_2".as("eventTime"))
+        .withWatermark("eventTime", "1 second")
+        .as[(String, Timestamp)]
+        .groupByKey(x => x._1)
+        .transformWithState(new RunningCountStatefulProcessorWithEventTimeTimer(),
+          TimeoutMode.EventTime(),
+          OutputMode.Update())
+
+      testStream(result, OutputMode.Update())(
+        StartStream(),
+        AddData(inputData, ("a",
+          Timestamp.valueOf("2023-08-01 00:00:00"))),
+
+        AddData(inputData, ("b",
+          Timestamp.valueOf("2023-08-02 00:00:00"))),
+        CheckNewAnswer(("a", "1"), ("a", "-1"), ("b", "1")),
+
+        AddData(inputData, ("b", Timestamp.valueOf("2023-08-03 00:00:00"))),
+        CheckNewAnswer(("b", "2")), // watermark: 3rd august, timer t1 should have fired,
+
+        AddData(inputData, ("b", Timestamp.valueOf("2023-08-04 00:00:00"))),
+        AddData(inputData, ("c", Timestamp.valueOf("2023-08-04 00:00:00"))),
+        CheckNewAnswer(("c", "-1"), ("c", "1")),
+        AddData(inputData, ("d", Timestamp.valueOf("2023-08-06 00:00:00"))),
+        CheckNewAnswer(("d", "1")),
+        StopStream
+      )
+    }
+  }
+
+  test("transformWithState - streaming with rocksdb and event time timer " +
+    "and add/remove timers should succeed") {
+    withSQLConf(SQLConf.STATE_STORE_PROVIDER_CLASS.key ->
+      classOf[RocksDBStateStoreProvider].getName) {
+      val clock = new StreamManualClock
+
+      import java.sql.Timestamp
+      val inputData = MemoryStream[(String, Timestamp)]
+      val result = inputData.toDS()
+        .select($"_1".as("value"), $"_2".as("eventTime"))
+        .withWatermark("eventTime", "1 second")
+        .as[(String, Timestamp)]
+        .groupByKey(x => x._1)
+        .transformWithState(new RunningCountStatefulProcessorWithAddRemoveEventTimeTimer(),
+          TimeoutMode.EventTime(),
+          OutputMode.Update())
+
+
+      testStream(result, OutputMode.Update())(
+        StartStream(),
+        AddData(inputData, ("a", Timestamp.valueOf("2023-08-01 00:00:00"))),
+        CheckNewAnswer(("a", "1"), ("a", "-1")),
+
+        AddData(inputData, ("a", Timestamp.valueOf("2023-08-02 00:00:00"))),
+        CheckNewAnswer(("a", "2"), ("a", "-1")),
```

Comment:
Can we also check the watermark stats here - something like https://src.dev.databricks.com/databricks/runtime@86b1eaff9fbdd28a8295b8201487317628677a5f/-/blob/sql/core/src/test/scala/org/apache/spark/sql/streaming/EventTimeWatermarkSuite.scala?L1063:15-1063:31